File: squareRoot/program/python/benchmark.py
"""Eksperyment: pomiar czasu miliona obliczen pierwiastka kwadratowego.

Porownuje trzy sposoby policzenia pierwiastkow z liczb 1..1 000 000:

  sqrt         - funkcja biblioteczna math.sqrt
  heron_naive  - metoda babilonska, punkt startowy x0 = n
  heron_fast   - metoda babilonska, punkt startowy z polowienia wykladnika

Uruchomienie:

    python benchmark.py                      # 5 prob po milion
    python benchmark.py --trials 3 --max 200000

Wyniki ladują w ../../wyniki/python-czasy.csv i python-srodowisko.csv.

DWIE PULAPKI POMIAROWE, ktore ten program omija:

1. Martwy kod. Kazda petla akumuluje sume wynikow i program ja wypisuje.
   Bez tego interpreter (a w wersji C++ optymalizator) moglby pominac
   obliczenia, ktorych wynik nie jest nigdzie uzywany, i pomiar stalby sie
   bezwartosciowy. Sumy z trzech metod musza byc zgodne - jesli nie sa,
   to znaczy, ze metody licza rozne rzeczy i porownanie nie ma sensu.

2. Zimny procesor. Kolejnosc metod jest rotowana miedzy probami. Gdyby byla
   stala, pierwsza mierzona metoda zawsze trafialaby na procesor przed
   podniesieniem taktowania i bylaby systematycznie karana.
"""

import argparse
import csv
import math
import platform
import sys
import time
from pathlib import Path

from sqrt_metody import heron_fast, heron_naive

WYNIKI = Path(__file__).resolve().parents[2] / "wyniki"


# Trzy osobne funkcje mierzace zamiast jednej z parametrem - z tego samego
# powodu, dla ktorego sqrt_metody ma dwie osobne petle Herona. Wywolanie
# przez zmienna kosztuje w kazdym z miliona obrotow.

def zmierz_sqrt(limit):
    sqrt = math.sqrt
    suma = 0.0
    start = time.perf_counter()
    for n in range(1, limit + 1):
        suma += sqrt(n)
    return time.perf_counter() - start, suma


def zmierz_naiwny(limit):
    fn = heron_naive
    suma = 0.0
    start = time.perf_counter()
    for n in range(1, limit + 1):
        suma += fn(n)
    return time.perf_counter() - start, suma


def zmierz_szybki(limit):
    fn = heron_fast
    suma = 0.0
    start = time.perf_counter()
    for n in range(1, limit + 1):
        suma += fn(n)
    return time.perf_counter() - start, suma


METODY = [
    ("sqrt", zmierz_sqrt),
    ("heron_naive", zmierz_naiwny),
    ("heron_fast", zmierz_szybki),
]


def nazwa_procesora():
    """Rzeczywisty model procesora.

    platform.processor() na Windowsie zwraca tylko ogolny opis rodziny
    ("Intel64 Family 6 Model 158..."), ktory nie identyfikuje maszyny.
    Konkretny model jest w rejestrze.
    """
    if platform.system() == "Windows":
        try:
            import winreg

            klucz = winreg.OpenKey(
                winreg.HKEY_LOCAL_MACHINE,
                r"HARDWARE\DESCRIPTION\System\CentralProcessor\0",
            )
            with klucz:
                return winreg.QueryValueEx(klucz, "ProcessorNameString")[0].strip()
        except OSError:
            pass
    return platform.processor() or "nieznany"


def nazwa_systemu():
    """Nazwa systemu z uwzglednieniem Windows 11.

    platform.release() zwraca "10" takze na Windows 11 - rozroznia je
    dopiero numer kompilacji (11 zaczyna sie od 22000).
    """
    system = platform.system()
    wydanie = platform.release()
    if system == "Windows" and wydanie == "10":
        try:
            build = int(platform.version().split(".")[-1])
            if build >= 22000:
                wydanie = "11"
        except (ValueError, IndexError):
            pass
    return f"{system} {wydanie} (build {platform.version()})"


def srodowisko():
    return {
        "system": nazwa_systemu(),
        "procesor": nazwa_procesora(),
        "python": platform.python_version(),
        "implementacja": platform.python_implementation(),
        "maszyna": platform.machine(),
    }


def przeprowadz(proby, limit):
    czasy = {nazwa: [] for nazwa, _ in METODY}
    sumy = {}
    for proba in range(1, proby + 1):
        przesuniecie = (proba - 1) % len(METODY)
        kolejnosc = METODY[przesuniecie:] + METODY[:przesuniecie]
        for nazwa, funkcja in kolejnosc:
            czas, suma = funkcja(limit)
            czasy[nazwa].append(czas)
            sumy[nazwa] = suma
            print(f"  proba {proba}  {nazwa:<12} {czas:10.4f} s")
    return czasy, sumy


def zapisz_czasy(czasy, proby):
    WYNIKI.mkdir(parents=True, exist_ok=True)
    sciezka = WYNIKI / "python-czasy.csv"
    with sciezka.open("w", newline="", encoding="utf-8") as plik:
        zapis = csv.writer(plik)
        zapis.writerow(["proba", "sqrt", "heron_naive", "heron_fast"])
        for i in range(proby):
            zapis.writerow(
                [i + 1] + [f"{czasy[nazwa][i]:.6f}" for nazwa, _ in METODY]
            )
        zapis.writerow(
            ["srednia"]
            + [f"{sum(czasy[nazwa]) / proby:.6f}" for nazwa, _ in METODY]
        )
    return sciezka


def zapisz_srodowisko(limit, proby, suma_kontrolna):
    WYNIKI.mkdir(parents=True, exist_ok=True)
    sciezka = WYNIKI / "python-srodowisko.csv"
    dane = srodowisko()
    dane["zakres"] = f"1..{limit}"
    dane["proby"] = proby
    dane["zegar"] = "time.perf_counter"
    dane["kolejnosc"] = "rotowana miedzy probami"
    dane["suma_kontrolna"] = f"{suma_kontrolna:.7f}"
    with sciezka.open("w", newline="", encoding="utf-8") as plik:
        zapis = csv.writer(plik)
        zapis.writerow(["klucz", "wartosc"])
        for klucz, wartosc in dane.items():
            zapis.writerow([klucz, wartosc])
    return sciezka


def main():
    parser = argparse.ArgumentParser(description="Pomiar czasu obliczania pierwiastkow.")
    parser.add_argument("--trials", type=int, default=5, help="liczba prob")
    parser.add_argument("--max", type=int, default=1_000_000, dest="limit",
                        help="gorna granica zakresu")
    args = parser.parse_args()

    print(f"Zakres 1..{args.limit}, liczba prob: {args.trials}")
    for klucz, wartosc in srodowisko().items():
        print(f"  {klucz}: {wartosc}")
    print()

    czasy, sumy = przeprowadz(args.trials, args.limit)

    print("\nSumy kontrolne (musza byc zgodne):")
    for nazwa, wartosc in sumy.items():
        print(f"  {nazwa:<12} {wartosc:.7f}")
    rozrzut = max(sumy.values()) - min(sumy.values())
    if rozrzut > 1e-3:
        print(f"\nBLAD: sumy kontrolne rozjezdzaja sie o {rozrzut}", file=sys.stderr)
        return 1

    print("\n| Proba | sqrt() | Heron (naiwny) | Heron (szybki) |")
    print("|---|---|---|---|")
    for i in range(args.trials):
        wiersz = " | ".join(f"{czasy[nazwa][i]:.4f} s" for nazwa, _ in METODY)
        print(f"| {i + 1} | {wiersz} |")
    srednie = " | ".join(
        f"{sum(czasy[nazwa]) / args.trials:.4f} s" for nazwa, _ in METODY
    )
    print(f"| Srednia | {srednie} |")

    plik_czasy = zapisz_czasy(czasy, args.trials)
    plik_srodowisko = zapisz_srodowisko(args.limit, args.trials, sumy["sqrt"])
    print(f"\nZapisano: {plik_czasy}")
    print(f"Zapisano: {plik_srodowisko}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: squareRoot/program/cpp/benchmark.cpp
// Eksperyment w C++: pomiar czasu miliona obliczen pierwiastka kwadratowego.
//
// Odpowiednik benchmark.py. Mierzy te same trzy metody, na tym samym
// zakresie, tym samym kryterium stopu - roznica jest wylacznie w jezyku.
//
// Kompilacja (flagi sa czescia metodyki, nie zmieniaj ich bez odnotowania):
//
//     g++ -O2 -std=c++17 -march=native -o benchmark.exe benchmark.cpp
//
// BARIERA OPTYMALIZATORA. Bez akumulacji sumy i bez bariery kompilator
// z flaga -O2 usuwa cala mierzona petle jako martwy kod, bo jej wynik nie
// jest nigdzie uzywany. Pomiar pokazalby wtedy 0 s - wynik bezwartosciowy.
// To jest ta sama pulapka, ktora w Pythonie omijamy suma kontrolna, tyle ze
// w jezyku kompilowanym jest znacznie grozniejsza.

#include <chrono>
#include <cstdio>
#include <cstdlib>
#include <vector>

#include "sqrt_metody.hpp"

namespace {

// Uniemozliwia optymalizatorowi usuniecie obliczen: mowi kompilatorowi,
// ze wartosc zostala odczytana i zmodyfikowana przez kod, ktorego nie widzi.
inline void nie_optymalizuj(double& wartosc) {
#if defined(__GNUC__)
    asm volatile("" : "+g"(wartosc) : : "memory");
#else
    volatile double sink = wartosc;
    (void)sink;
#endif
}

// steady_clock, a nie system_clock - ten drugi moze skoczyc w trakcie
// pomiaru, jesli system zsynchronizuje zegar z serwerem czasu.
using Zegar = std::chrono::steady_clock;

struct Wynik {
    double sekundy;
    double suma;
};

template <typename Fn>
Wynik zmierz(Fn fn, long limit) {
    double suma = 0.0;
    const auto start = Zegar::now();
    for (long n = 1; n <= limit; ++n) {
        suma += fn(static_cast<double>(n));
    }
    nie_optymalizuj(suma);
    const auto stop = Zegar::now();
    const std::chrono::duration<double> czas = stop - start;
    return {czas.count(), suma};
}

double srednia(const std::vector<double>& v) {
    double s = 0.0;
    for (double x : v) s += x;
    return s / static_cast<double>(v.size());
}

}  // namespace

int main(int argc, char** argv) {
    int proby = 5;
    long limit = 1000000;
    if (argc > 1) proby = std::atoi(argv[1]);
    if (argc > 2) limit = std::atol(argv[2]);

    std::printf("Zakres 1..%ld, liczba prob: %d\n\n", limit, proby);

    std::vector<double> t_sqrt, t_naive, t_fast;
    double s_sqrt = 0.0, s_naive = 0.0, s_fast = 0.0;

    for (int p = 1; p <= proby; ++p) {
        // Rotacja kolejnosci metod miedzy probami - tak samo jak w wersji
        // pythonowej, zeby zadna metoda nie byla systematycznie karana
        // za trafienie na procesor przed podniesieniem taktowania.
        const int przesuniecie = (p - 1) % 3;
        Wynik a{}, b{}, c{};
        for (int krok = 0; krok < 3; ++krok) {
            switch ((przesuniecie + krok) % 3) {
                case 0: a = zmierz([](double x) { return std::sqrt(x); }, limit); break;
                case 1: b = zmierz(pierwiastek::heron_naive, limit); break;
                case 2: c = zmierz(pierwiastek::heron_fast, limit); break;
            }
        }
        t_sqrt.push_back(a.sekundy);
        t_naive.push_back(b.sekundy);
        t_fast.push_back(c.sekundy);
        s_sqrt = a.suma; s_naive = b.suma; s_fast = c.suma;
        std::printf("  proba %d  sqrt %8.4f s  naiwny %8.4f s  szybki %8.4f s\n",
                    p, a.sekundy, b.sekundy, c.sekundy);
    }

    std::printf("\nSumy kontrolne (musza byc zgodne):\n");
    std::printf("  sqrt         %.7f\n", s_sqrt);
    std::printf("  heron_naive  %.7f\n", s_naive);
    std::printf("  heron_fast   %.7f\n", s_fast);

    FILE* f = std::fopen("../../wyniki/cpp-czasy.csv", "w");
    if (!f) {
        std::fprintf(stderr, "BLAD: nie mozna otworzyc pliku wynikowego\n");
        return 1;
    }
    std::fprintf(f, "proba,sqrt,heron_naive,heron_fast\n");
    for (int i = 0; i < proby; ++i) {
        std::fprintf(f, "%d,%.6f,%.6f,%.6f\n",
                     i + 1, t_sqrt[i], t_naive[i], t_fast[i]);
    }
    std::fprintf(f, "srednia,%.6f,%.6f,%.6f\n",
                 srednia(t_sqrt), srednia(t_naive), srednia(t_fast));
    std::fclose(f);

    std::printf("\n| Proba | sqrt() | Heron (naiwny) | Heron (szybki) |\n");
    std::printf("|---|---|---|---|\n");
    for (int i = 0; i < proby; ++i) {
        std::printf("| %d | %.4f s | %.4f s | %.4f s |\n",
                    i + 1, t_sqrt[i], t_naive[i], t_fast[i]);
    }
    std::printf("| Srednia | %.4f s | %.4f s | %.4f s |\n",
                srednia(t_sqrt), srednia(t_naive), srednia(t_fast));
    std::printf("\nZapisano: ../../wyniki/cpp-czasy.csv\n");
    return 0;
}
